torch_xla/debug/profiler.py

```diff
@@ -68,3 +68,50 @@ def trace(service_addr: str,
       duration_ms=duration_ms,
       num_tracing_attempts=num_tracing_attempts,
       options=options)
+
+
+class Trace(torch_xla._XLAC.profiler.TraceMe):
+  """Context manager that produces a trace event for profiling.
+
+  The traces generated can then be collected using the above profiling APIs.
+  The profiling server first needs to be started up and then can be sampled
+  either using Tensorboard profiler plugin
+  (https://github.com/tensorflow/profiler) or the
+  :func:`~torch_xla.debug.profiler.trace` method.
+
+  Note: currently only supports PyTorch/XLA client side trace events. i.e.,
+  the namespace won't group TPU worker side trace.
+
+  Example usage:
+  ```python
+  server = xp.start_server(9012)
+
+  with xp.Trace('fwd_context'):
+    model(input)
+    xm.mark_step()
+  ```
+  """
+  pass
+
+
+class StepTrace(torch_xla._XLAC.profiler.TraceMe):
```

Comment:
Why do we separate this class? Can't we use `Trace` at all times?

Reply:
It's a bit of magic on xprof's part but basically when it comes to performance analysis on xprof, they depend on us to provide explicit step markers as "root". This is one by `_r=1` *kwarg: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/profiler/utils/xplane_schema.cc#L169

Reply:
We can have one single `Trace` class, and mark_step by default. We can then use it as:

```python
with xp.Trace('trainstep', step_num=i):
  with xp.Trace('graphbuild', markstep=False):
    ...
```

Or we can have a `StepTrace` class that is equivalent to `Trace` with always `markstep=True`. Imho the first is cleaner. We're forcing users to know the difference b/w marking and not marking the step anyway. wdyt?

Reply:
I personally think that it might be a good idea to keep the `TraceStep` pretty explicit given the large behavior difference between the two when it comes to tracing/marking. If you have strong opinions I can change.

Reply:
yeah ok let's keep it different.